staging/src/k8s.io/apiserver/pkg/util/flowcontrol/dropped_requests_tracker.go

```diff
@@ -0,0 +1,184 @@
+	// RetryAfter value.
+	GetRetryAfter(plName string) int64
+}
+
+// unixStat keeps a statistic how many requests were dropped within
+// a single second.
+type unixStat struct {
+	unixSecond int64
+	requests   int64
+}
+
+type droppedRequestsStats struct {
+	history []unixStat
+
+	currentRetryAfterStep int
+	retryAfterUpdateUnix  int64
+}
+
+func (s *droppedRequestsStats) recordDroppedRequest(unixTime int64) {
+	if length := len(s.history); length == 0 || s.history[length-1].unixSecond != unixTime {
+		s.history = append(s.history, unixStat{unixSecond: unixTime, requests: 0})
+	}
+	s.history[len(s.history)-1].requests++
+
+	s.trimHistory(unixTime)
+}
+
+func (s *droppedRequestsStats) trimHistory(unixTime int64) {
+	startIndex := 0
+	for ; startIndex < len(s.history) && unixTime - s.history[startIndex].unixSecond >= maxHistory; startIndex++ {}
+	if startIndex > 0 {
+		s.history = s.history[startIndex:]
+	}
+}
+
+// recomputeRetryAfter is checking if retryAfter shouldn't be adjusted.
+// We're using the following logic:
+// - if retryAfter was adjusted less than `retryAfter` seconds ago, skip
+// - compute how many requests were dropped in the last `retryAfter` seconds
+// - given that APF is queuing requests, and rejecting them as a last resort
+//   we expect this value to be fairly low, unless the system is actually
+//   overloaded; as a result, we arbitrarily say that if the computed number
+//   of dropped requests is larger than `3*retryAfter`, we increase the
+// - similarly, we decrease the retryAfter value only of over the recent
+//   `retryAfter` seconds period, it didn't exceed `retryAfter` dropped
+//   requests.
+func (s *droppedRequestsStats) recomputeRetryAfter(unixTime int64) {
+	retryAfter := retryAfterSteps[s.currentRetryAfterStep]
+
+	if unixTime - s.retryAfterUpdateUnix >= retryAfter && s.currentRetryAfterStep < len(retryAfterSteps) - 1 {
+		// Check if we're meeting the condition to increase.
+
+		// TODO(wojtek-t): Precompute this value.
+		droppedRequests := int64(0)
+		for _, h := range s.history {
+			if unixTime - h.unixSecond < retryAfter {
+				droppedRequests += h.requests
+			}
+		}
+		if droppedRequests >= 3 * retryAfter {
+			s.currentRetryAfterStep++
+			s.retryAfterUpdateUnix = unixTime
+			return
+		}
+	}
+
+	if unixTime - s.retryAfterUpdateUnix > retryAfter && s.currentRetryAfterStep > 0 {
+		// Check if we're meeting the condition to decrease.
+
+		// TODO(wojtek-t): Precompute this value.
+		droppedRequests := int64(0)
+		for _, h := range s.history {
+			if unixTime - h.unixSecond <= retryAfter {
+				droppedRequests += h.requests
+			}
+		}
+		if droppedRequests < retryAfter {
+			s.currentRetryAfterStep--
+			s.retryAfterUpdateUnix = unixTime
+			return
+		}
+	}
+}
+
+// droppedRequestsTracker implement DroppedRequestsTracker interface
+// for the purpose of adjusting RetryAfter header for newly dropped
+// requests to avoid system overload.
+type droppedRequestsTracker struct {
+	clock clock.Clock
```

Comment:
You don't need a whole Clock here, a PassiveClock or `func() time.Time` will do.

Reply:
done